### DISCUSSION TITLE
stutter when dynamically adding/removing navmeshes

### DISCUSSION THREAD

**Discussion by GGAlanSmithee** [isaac-mason/recast-navigation-js#405] [Q&A] [answered]
In my game, I am doing what I think is a rather common thing, which is that I have split my world into chunks and am dynamically loading/unloading them as the user approaches/moves away from a zone.

For now, I just created the most naive solution possible, which is that I conditionally render the zones as children of my navmesh (R3F) and use a `useEfect` hook to rebuild the navmesh. However, this causes a visible stutter every time a user approaches a zone, which is understandable.

I haven't looked into it yet, since I wanted to check if there are known solutions to this. One thing I am thinking about is using the tilecache, like I am with obstacles, but it doesn't "feel" like that's a correct solution, since these chunks aren't obstacles.

Thanks in advance!

**Comment by GGAlanSmithee** [comment]
So, I used the example provided here https://github.com/isaac-mason/recast-navigation-js/tree/next/examples/vite-worker-example, using a worker, and it works great!

The time to generate a navmesh are somewhat longer, but comparable, but since they are done in a worker thread, there is no visible stutter at all! Very nice.

However, when attempting to use a tilecache (`generateTileCache`) I get an error (`Uncaught RuntimeError: null function or function signature mismatch`) when later calling `importNavMesh`, and my guess is that I need to pass a `TileCacheMeshProcess`, but not quite sure how to construct this.

Thanks again!

**Comment by GGAlanSmithee** [comment]
I dug up this from the source: https://github.com/isaac-mason/recast-navigation-js/blob/main/packages/recast-navigation-generators/src/generators/generate-tile-cache.ts#L117C1-L123C6

```ts
const tcmp = new TileCacheMeshProcess((navMeshCreateParams, polyAreas, polyFlags) => {
        for (let i = 0; i < navMeshCreateParams.polyCount(); ++i) {
          polyAreas.set(i, 0)
          polyFlags.set(i, 1)
        }
      })
```

Now, everything is working great 🎉 thanks for this amazing package! 🙏🏼

**Comment by GGAlanSmithee** [comment]
Here's the result btw


https://github.com/isaac-mason/recast-navigation-js/assets/11805103/4b25c41d-0ca8-4c1d-b1dd-b18b1507bfc5

**Comment by isaac-mason** [comment]
Nice! Thanks for sharing 😁 

It's also possible to build just the nav mesh tiles on a worker, and maintain one nav mesh on the main thread / another thread.

This is what this demo does:
https://sketches.isaacmason.com/sketch/recast-navigation/dynamic-tiled-navmesh
https://github.com/isaac-mason/sketches/blob/main/src/sketches/recast-navigation/dynamic-tiled-navmesh/navigation/dynamic-tiled-navmesh.ts

This is not as straightforward to do with a TileCache right now though. In the c++ library itself the DetourTileCache takes more control over nav mesh building so it's less flexible, and also this library generally needs to expose more of the TileCache interface.

**Comment by isaac-mason** [comment]
There's probably also an action here to improve the docs for importing/exporting a TileCache :) 

I've also considered having a seperate `importTileCache` / `exportTileCache` so the import/export api less confusing, I might pick that back up.

**Comment by GGAlanSmithee** [comment]
@isaac-mason I think the implementation as it is now is pretty straight forward, once you get it, with the exception of `TileCacheMeshProcess` which I still don't really understand, as I just copy + pasted it :) I think adding some documentation on this would go a long way though!

**Reply by isaac-mason** [reply, marked as answer]
@GGAlanSmithee I've ended up splitting importNavMesh/exportNavMesh and importTileCache/exportTileCache as I've realised importNavMesh was unnecessarily creating tilecache structures when importing just a navmesh.

I've also added to the import/export tilecache readme section so hopefully the usage is clearer 🙂

**Reply by isaac-mason** [reply]
I'll add some more general documentation on TileCacheMeshProcess too :)

**Reply by GGAlanSmithee** [reply]
@isaac-mason amazing! You rock 🥇
